### PR TITLE
[reland] Fix estimate_nccl_collective_runtime

### PR DESCRIPTION
**Pull request by yifuwang**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #119229
* #119104
* #119103
* __->__ #118986
* #119102

`estimate_nccl_collective_runtime` has been broken and the errors have been silently swallowed by inductor. This PR:
- Fixes the issues described in https://github.com/pytorch/pytorch/issues/118497.
- Adds white-box testing so future issues can be surfaced in tests.
- Add support for native funcol IRs.

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225 @voznesenskym @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler

### CODE REVIEW

File: test/inductor/test_snode_runtime.py [hunk 1]
```diff
@@ -165,6 +172,132 @@ def f(x):
...
+    def _verify_runtime_estimation(self, fn, inps):
+        store = FakeStore()
+        dist.init_process_group(
+            backend="fake", rank=0, world_size=self.WORLD_SIZE, store=store
+        )
+        try:
+            metrics.reset()
+            torch._dynamo.optimize(count_bytes_inductor)(fn)(*inps)
+            found_collective = False
+            for snode, runtime in metrics.node_runtimes:
```
Comment (wconstab) [inline test/inductor/test_snode_runtime.py:188]: cool, is this an estimate from the scheduler for every node, (iirc there were some lo-fi estimates like eltwise was cheap and gemm is expensive, but I didn't remember if it was calibrated (is the collective estimator predicting actual time intervals?)

File: test/inductor/test_snode_runtime.py [hunk 2]
```diff
@@ -165,6 +172,132 @@ def f(x):
...
+        def fn(x):
+            rs = _c10d.all_reduce_coalesced(x, "sum", "0")
+            return [_c10d.wait_tensor(r) for r in rs]
+
+        inp = [T(10, 10), T(15, 15)]
+        self._verify_runtime_estimation(fn, (inp,))
+
+    def test_all_gather_into_tensor(self):
+        def fn(x):
+            rs = _c10d.all_gather_into_tensor(
```
Comment (wconstab) [inline test/inductor/test_snode_runtime.py:252]: hm, what's this doing- these are not functional collectives, so I wouldn't have expected them to work with wait_tensor.  or, is that counting on dynamo doing a rewrite?

File: torch/_inductor/scheduler.py [hunk 3]
```diff
@@ -581,6 +583,16 @@ def get_estimated_runtime(self) -> float:
             # default to no reordering based on runtime
             return 0
 
+        # Collective kernels
```
Comment (wconstab) [inline torch/_inductor/scheduler.py:586]: you had to move this higher up since it was getting bypassed previously?

File: test/inductor/test_snode_runtime.py (hunk 1 above)
Comment (wconstab) [reply test/inductor/test_snode_runtime.py:188]: nvm see it below. time in ns.

File: torch/_inductor/utils.py [hunk 4]
```diff
@@ -1258,3 +1258,15 @@ def pass_execution_and_save(func, gm, msg):
             t,
             time_elapsed,
         )
+
+
+def is_collective(node):
+    from . import ir
+
+    return isinstance(node, ir.CollectiveKernel) or type(node) == ir._CollectiveKernel
```
Comment (wconstab) [inline torch/_inductor/utils.py:1266]: this is subtle, something related to new types of collectives being lowered differently?

File: test/inductor/test_snode_runtime.py (hunk 2 above)
Comment (yifuwang) [reply test/inductor/test_snode_runtime.py:252]: Hey @wconstab, these are dispatcher ops. For both the existing and new funcol, "tensor wrapping" happens inside dispatcher calls.

The reason I resorted to dispatcher ops for this test is that there's currently some issues with tracing some of the new funccols. This is okay for this test since it focuses on the IR. I will add tracing tests as I resolve the aforementioned issues.

File: torch/_inductor/utils.py (hunk 4 above)
Comment (yifuwang) [reply torch/_inductor/utils.py:1266]: You mean `type(node) == ir._CollectiveKernel`? In the new IRs, `ir._WaitKernel` derives from `ir._CollectiveKernel` to share some stuff needed for the cpp codegen.

File: torch/_inductor/scheduler.py (hunk 3 above)
Comment (yifuwang) [reply torch/_inductor/scheduler.py:586]: Yes. Previously this part always failed and returned 0 because for collective kernels, the dtype obtained above is always `None`.

```python
        try:
            gpu_memory_bandwidth = get_gpu_dram_gbps()
            gpu_flops = get_device_tflops(dtype) * 10**12
        except Exception:
            return 0
```

File: torch/csrc/distributed/c10d/FakeProcessGroup.hpp [hunk 5]
```diff
@@ -9,6 +9,12 @@ class FakeWork : public Work {
   bool wait(std::chrono::milliseconds timeout) override {
     return true;
   }
+
+  c10::intrusive_ptr<c10::ivalue::Future> getFuture() override {
+    auto fut = c10::make_intrusive<c10::ivalue::Future>(c10::NoneType::get());
+    fut->markCompleted();
+    return fut;
+  }
```
Comment (yf225) [inline torch/csrc/distributed/c10d/FakeProcessGroup.hpp:17]: dumb q: is this part related to the PR?

**Review by yf225 — APPROVED** [review]
Thanks a lot for fixing this @yifuwang !

**Comment by kwen2501** [comment]
Hi @yifuwang sorry just get to notice that we have this utility. (Which would be indeed helpful for comm scheduling)
I wonder if there would be enough time for pytorch to read in a matrix dumped by NCCL during init?
I previously opened a PR towards NCCL to dump its tuning matrix containing the bandwidth and latency numbers. See https://github.com/NVIDIA/nccl/issues/830.
Since it is prepared **AFTER** NCCL performs topology detection, its bandwidth/latency numbers are most accurate.
In PyTorch, I'd think we would avoid any topology detection (too low level, also lack of APIs). 
Also, the tuning matrix of NCCL would match how NCCL selects algorithm and protocol on the specific platform the job is run on.
